### ISSUE TITLE
Use nested options for PhET-iO options

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/tandem#230]
In a Slack discussion where I was grousing about verbose option names for PhET-iO, @samreid asked if I had any recommendations.   Glad you asked!

Options related to PhET-iO are currently "flat".  A big improvement would be to use nested options, putting all PhET-iO options under `phetioOptions`. 

Benefits of using nested options here:
* It conforms to PhET standards for options and nested options.  This seems like an ideal (and very important) place to follow those standards, and I can think of no good reason not to follow them.
* It forces related options to be grouped together, so you can look in 1 place to find options that are specific to phetio.  Grouping is currently a convention that is not strictly followed or enforced.  
* The current convention (again, not always followed or enforced) is to manually group all phetio options at the end of the option hash, preceeded by a `// phet-io` comment.  If we nest `phetioOptions`, we can drop that (somewhat onerous) requirement.
* It eliminates the need to put "phetio" in option names, resulting in less-verbose names.
* If you aren't familiar with an option, it is easily identified as being related to phetio. 

Compare:

```js
// flat options
options = merge( {
  phetioDocumentation: 'blah blah blah'
  tandem: Tandem.REQUIRED,
  phetioState: false,
  fill: 'red',
  stroke: 'yellow',
  phetioType: BooleanIO,
  lineWidth: 2,
  phetioReadOnly: true,
  inputEnabledPropertyPhetioInstrumented: true,
  phetioStudioControl: false
}, options );

// nested options
options = merge( {
  fill: 'red',
  stroke: 'yellow',
  lineWidth: 2,
  phetioOptions: {
    tandem: Tandem.REQUIRED,
    ioType: BooleanIO,
    inputEnabledPropertyInstrumented: true,
    documentation: 'blah blah blah',
    readOnly: true,
    state: false,
    studioControl: false,
  }
}, options );
```

This seems like a no-brainer to me. Yes, there would be a significant cost to change existing code.  But I think it's totally worth it.

I used `phetioOptions` above, because it replaces the `// phet-io` grouping of options.  But I don't feel strongly about that name, and would be happy for the PhET-iO team to choose a better name.

Nested options should also be applied to PDOM (a11y) options, for all of the same reasons. See https://github.com/phetsims/scenery/issues/1190.

I'm going to give this a high priority for discussion. If this is going to be done, it's advantageous to do it before more code is created that will need to be changed.

**Comment by pixelzoom** [comment]
Here's a plan for converting to `phetioOptions`:

1. Identify all of the options that are currently specified in the "flat" `options` hash.
2. Where those options are used, convert to `phetOptions` nested options. Add assertions that the flat options are undefined in `options`, to catch any call sites that might be missed.
3. Convert call sites to `phetioOptions`.
4. Test locally in aqua to identify call sites that were missed. 
5. Remove the assertions that were added in 2.

**Comment by zepumph** [comment]
In general I am a bit more weary of this pattern than you seem. I don't want to diminish the original impetus for this issue, in which there seems to be a burden to options management because of PhET-iO naming convention's verbosity. I want to work on that as far as we as a project feel like it needs to. That said I have some thoughts creating `phetioOptions` nesting.

*>It conforms to PhET standards for options and nested options.

I thought nested options was a pattern used for composition, and passing options to sub components. Will you please provide a spot where a nesting occurs for a supertype's options? Most of these phet-io options are provided to `PhetioObject`, so that is basically what we are doing. Would we want `nodeOptions` to organize all the options that you pass up to Node?


* Your flat options example does not adhere to our code standards. It doesn't pull out the phet-io options into its own block. So the comparison is, at least slightly, skewed in favor of a strategy that groups these options together. That said I hear your point above about this being a manual separation.
* (Chief complaint) There is a burden to adding support for a semi arbitrary grouping of `phetioOptions` to all places that include an option specific to their phet-io brand. For example, `Checkbox.phetioLinkProperty`  is defined and used only in Checkbox. Should we say that this now needs to be provided via a separate options block just because of the brand that it is used in? It would be the only one that Checkbox supports. That feels off to me. I don't really understand that. The same is true in your example above for Node.inputEnabledPropertyPhetioInstrumented (which is an option for Node). Basically this will great a cross product of possibilities when searching for how to provide options to a super type. We will need to teach ourselves to look at the super type flat level options, and also for a specific nested options block `phetioOptions` to apply to that very component.
* We began prefixing options with `phetio` to decrease the burden of understanding for this whole new feature set that we retrofitted into the project. I fail to see how this grouping of options is any different from other examples of types that use a lot of options. PhetioObject has a bunch of options. It would be more confusing to me to nest those just because we decided that they should all start with phetio. If verbosity is the problem, first I would think about recognizing the growth in our team's understanding, and perhaps no longer needing the "training wheels" of `phetio` prefixed options. They are just options, like any other Type. A semi-arbitrary classification based on brand does seem helpful on a component level. If `phetioOptions` gets traction, I would wonder why we wouldn't also have `layoutOptions`, as there are many of those, potentially influencing many types up and down certain Node hierarchies.

I want to go back to the problem at hand, which from the slack discussion seems to be "the option names are too long". I'm happy to shorten some of these if that would be helpful. For example the Node convention of `xPropertyPhetioInstrumented` seems like it could use a shortening. In the majority of cases, where there are just a couple of options, this will likely be more verbose:

```
x = {
  phetioOptions: {
    state: true,
    readOnly: true
  }
}


x = {
  phetioState: true,
  phetioReadOnly: true
}
```

**Comment by pixelzoom** [comment]
> I thought nested options was a pattern used for composition, and passing options to sub components. 

If you consider nest options only relevant to nested compositions, that is a narrow (and imo incorrect) view of nested options. Yes, PhET has been using nested options mainly for composition, e.g. `buttonOptions` or `textOptions`.  (And I have concerns about that, because it breaks encapsulation by exposing implementation details, but I digress.)  There is absolutely no reason why nested options can't be used to group related options, with all of the same benefits.  And there is absolutely nothing about nested options (in the pattern or in our discussions) that excluses its use here. 

> There is a burden to adding support for a semi arbitrary grouping of phetioOptions to all places that include an option specific to their phet-io brand. 

I don't see any problem with adding class-specific options to `phetioOptions` when they are related to PhET-iO.  How is `options.phetioOptions.linkProperty` more burdensome than `options.phetioLinkProperty`??  What's burdensome is having to manually scan through all of the `options`, and manaully determine what may and may not be related to PhET-iO.

> I want to go back to the problem at hand, which from the slack discussion seems to be "the option names are too long". 

That is not the problem at hand. That is the spark that motivated this issue and phetsims/scenery#1190, and frankly the verbose names is the least of the problems.  The problem is:

* A large and growing number of related options.
* No API support for grouping those options.
* A manual convention for grouping options - which is onerous, unenforceable, and frequently violated.
* Inabiltiy to recognize unfamiliar options because they are not grouped.
* Verbose options names because they need to be understandable in a "flat" options context.

**Comment by pixelzoom** [comment]
> We began prefixing options with phetio to decrease the burden of understanding for this whole new feature set that we retrofitted into the project. I fail to see how this grouping of options is any different from other examples of types that use a lot of options.   ...  

The basis for your arguement seems to be that this doesn't fit into the pattern of using flat options for class hierachy and using nested optons for composition.  Imo that's a weak argument, and a rather limited view of organizing options.  If there are advantages to be had to organizing in other ways (in this case, by feature set), I'd like to hear why that's not appropriate.

> ... If phetioOptions gets traction, I would wonder why we wouldn't also have layoutOptions, as there are many of those, potentially influencing many types up and down certain Node hierarchies.

Two big differences :

(1) PhET-iO is a huge cross-cutting aspect that impacts the entire PhET code base. Things like the options for LayoutBox (your example) are not cross cutting, and have nowhere near the impact.

(2) I'm running in this constantly for the PhET-iO feature set.  I have no problem with flat options for the typical class hierarchy case.

**Comment by samreid** [comment]
To understand the tradeoffs of the proposal in a practical scenario, I navigated to Fourier, searched for `phetioType` and converted the first case I saw, here is a picture of the diff:

![image](https://user-images.githubusercontent.com/679486/112909811-5d10f100-90af-11eb-8fd8-8052502cc68b.png)

I understand the argument for combining conceptually related options together, but the increased verbosity does not feel to me as if it "pulls its weight" in terms of improved clarity and maintainability.  This is a subjective call and I understand if there is disagreement about this.

This search next took me to ContinuousModel, where I saw instrumented Properties.  I tried converting them to the proposed approach:

![image](https://user-images.githubusercontent.com/679486/112909971-a9f4c780-90af-11eb-9513-4c21c5089cbd.png)

For cases like the one in the issue description, nesting the options seems reasonable, but I think there would be an abundance of cases where it feels like undesirable verbosity that does not add significant value.

> PhET-iO is a huge cross-cutting feature that impacts the entire PhET code base, and its options sometimes apply to things outside of class hierarchy or composition. 

I would not be surprised to see an example of this, but as far as I can tell, the options that are proposed to move to `phetioObjects` are all options for base class `PhetioObject`.

**Comment by pixelzoom** [comment]
I'm happy to just close this issue, if nesting PhET-iO options is going to create more problems than it addresses.

**Comment by samreid** [comment]
I'd appreciate a subgroup meeting between @samreid @pixelzoom and @zepumph to make sure I understand the issue.  It may even be a brief meeting where we ultimately close the issue.  But I don't think this needs to be marked for phet-io meeting.  Please correct me if I'm wrong about any of that.

**Comment by samreid** [comment]
I'm also wondering how we would solve this problem with typescript, and how our options pattern may be handled in typescript.  See https://github.com/phetsims/tasks/issues/987#issuecomment-820730972.  I'll reach out to @zepumph and @pixelzoom for scheduling a discussion

**Comment by pixelzoom** [comment]
Zoom meeting with @samreid @zepumph @pixelzoom 

We decided that PhET-iO options are fine as is, the "phetio" prefix is useful, and verbose option names are OK.  There are a couple of options that need a reminder about the "phetio" prefix (e.g. `enabledPropertyPhetioInstrumented`), but they will be addressed in phetsims/scenery#1202. 

Also... The "phetio" prefix indicates an API feature, does not indicate composition or inheritance. As such, clients are free to use "phetio" prefix for their own options.

Closing.
